buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/UpdateBranchStatus.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.gradlebuild
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.*
+import org.gradle.plugins.performance.determineCurrentBranch
+
+
+/**
+ * When a Ready for Nightly trigger build completes successfully on master/release, we publish a git tag
+ * green-master/green-release to remote repository so that developers can checkout new branches from these tags.
+ */
+open class UpdateBranchStatus : DefaultTask() {
+    @TaskAction
+    fun publishBranchStatus() {
+        when (project.determineCurrentBranch()) {
+            "master" -> publishBranchStatus("master")
+            "release" -> publishBranchStatus("release")
+        }
+    }
+
+    private
+    fun publishBranchStatus(branch: String) {
+        project.execAndGetOutput("git", "tag", "-d", "green-$branch")
```

Comment:
I think tag is more lightweight than a branch because adding/removing tag doesn't need to change any files, right?

Reply:
I know branch is a pointer to commit, just like a tag. By saying it "not lightweight", I mean when you create a new branch and push it, you need to `checkout -b green-branch` then `checkout -` back because some of our build logic depends on the branch name. Plus, when checking out branches, the files are possibly changed (though this shouldn't happen in our build). That's why I prefer tags.

Reply:
I always felt that using a tag is more correct. We want to "tag" the correct commit, not create a "branch" (which indicates a forking of timelines).

Reply:
Yeah, I also think tag is better than a branch.

Reply:
I don't like using tag, since a tag points to a fixed commit which cannot be changed. Therefore, we need to delete the tag and recreate it. That feels wrong to me.

Imagine Teamcity as a developer which keeps fast-forwarding the green branch when it verified the changes are complete.

Having said that, I don't feel very strongly about whether to use a tag or a branch. The (not-important) advantage of a branch is that we only need to use one git command instead of three. We will see in daily development whether a branch or a tag is more convenient.

Reply:
I was about to say that both tags and branches are wrong in some sense, but then I found that Git itself recognizes the need for temporary tags:

> It’s generally recommended that you create annotated tags so you can have all this information; but if you want a temporary tag or for some reason don’t want to keep the other information, lightweight tags are available too.

in https://git-scm.com/book/en/v2/Git-Basics-Tagging

So I'd say let's go for a tag.

For what use case do you need three commands instead of one when using tags? I can do `git switch -c lptr/test green-master` when `green-master` is a tag.

There's also the added advantage of using a tag that when I'm on `master`, I can see if the current commit is considered green or not on my command-line directly:

![image](https://user-images.githubusercontent.com/495366/66563051-3bd7e000-eb5d-11e9-85af-fdd8cc58d469.png)

Reply:
> For what use case do you need three commands instead of one when using tags?
When creating and pushing the tag, the tag needs to be deleted, created and then pushed. For a branch, you can directly push (i.e. one command).

After thinking some more, for me, the main advantage of using a tag is that nobody can push commits on top of the tag accidentally. If we use a branch, then it any developer could accidentally push more changes to it.

Let's go for a tag.

Reply:
After trying it out, the disadvantage of tag emerges:

- `git pull` won't update tag from remote. You have to use `git fetch --tags`.
- `git fetch --tags` will reject updating locally existing tag, you have to use `git fetch --tags -f`.

So now I'm inclined to use branch (so that you can update the `green-master` with a regular `git pull`). Regarding that developers might accidentally push commits to it, is there any other way to forbid direct push?

Reply:
Hm. Does this happen with lightweight tags or only with annotated ones?

About the need for `--tags`:

> So if your tags are reacheable from the branches you are fetching, you don't have to include `--tags` by default.

https://stackoverflow.com/a/12542314/104894

I think having to use `fetch` is not necessarily a problem. If we can make it work via:

```shell
$ git fetch && git switch -c $newBranch green-master
```

...then I'd be happy.

Reply:
I think the blocker for a tag is this:
> git fetch --tags will reject updating locally existing tag, you have to use git fetch --tags -f.

I saw this when running `git fetch --tags`:
```
From github.com:gradle/gradle
 ! [rejected]                green-master -> green-master  (would clobber existing tag)
```

The tag is a light-weight tag. Temporary still means immutable I suppose...

Reply:
Yeah... you have to force update `git fetch --tags -f`....

Let's switch back to branch...

Reply:
Alright, I concede. :)